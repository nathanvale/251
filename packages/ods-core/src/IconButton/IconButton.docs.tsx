import { ComponentDocs } from "@origin-digital/ods-types";
import { docs as buttonDocs } from "../Button/Button.docs";
import { IconButtonProps } from "./IconButton";
import { generateIconButton } from "./IconButton.helper";

export const docs: ComponentDocs<IconButtonProps> = {
  category: "Interaction",
  componentName: "IconButton",
  description: "Renders a button with an icon inside",
  propDescriptions: {
    children: "A single icon from the @origin-digital/icons package or SvgIcon",
    color:
      "The color of the icon. If inside a text component you can set to inherit and the tone from the text will be used instead",
    noIconPadding:
      "Removes the visible padding around the icon but leaves the touchable area",
    href: buttonDocs.propDescriptions!.href,
    target: buttonDocs.propDescriptions!.target,
  },
  migrationGuide: false,
  examples: {
    default: {
      Code: () => generateIconButton(),
    },
    additional: [
      {
        label: "Inheriting text color",
        Code: () =>
          generateIconButton({ color: "inherit" }, {}, { tone: "positive" }),
      },
    ],
  },
  snippets: [
    {
      label: "Default",
      Code: () => generateIconButton(),
    },
  ],
};
