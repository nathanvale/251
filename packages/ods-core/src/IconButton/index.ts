export { IconButton } from "./IconButton";
export type IconButtonProps = import("./IconButton").IconButtonProps;
