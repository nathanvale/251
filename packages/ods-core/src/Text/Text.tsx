import React, { ReactNode, useMemo } from "react";
import clsx from "clsx";
import { BoxProps, Box } from "@origin-digital/ods-core";
import {
  TextVariants,
  TextToneVariants,
  TypographyWeightVariants,
} from "@origin-digital/ods-types";
import {
  useBasekickStyles,
  useToneStyles,
  useWeightStyles,
  useStrongStyles,
} from "@origin-digital/ods-typography";
import {
  useCheckTypographyBackground,
  useTruncatedContent,
} from "../_private/hooks";

import { TextContext } from "./TextContext";

export interface UseTextProps {
  mediumWeight?: boolean;
  variant?: TextVariants;
  tone?: TextToneVariants;
}

export interface TextProps {
  children?: ReactNode;
  variant?: TextVariants;
  tone?: TextToneVariants;
  weight?: Exclude<TypographyWeightVariants, "bold">;
  align?: BoxProps["textAlign"];
  truncate?: boolean;
  "data-id"?: string;
  component?: BoxProps["component"];
}

interface UseTextStylesProps {
  weight: Exclude<TypographyWeightVariants, "bold">;
  tone?: TextToneVariants;
  variant: TextVariants;
}

const defaultWeight = "regular";
const defaultVariant = "body";

function useTextStyles(props: UseTextStylesProps) {
  const { tone = "neutral", weight, variant } = props;
  useCheckTypographyBackground();
  const basekickStyles = useBasekickStyles({ variant });
  const toneStyles = useToneStyles({ tone });
  const weightStyles = useWeightStyles(weight);
  const strongStyles = useStrongStyles("medium");
  return clsx(basekickStyles, toneStyles, weightStyles, strongStyles);
}

export const Text = (props: TextProps) => {
  const {
    children,
    component = "span",
    variant = defaultVariant,
    truncate = false,
    align,
    weight = defaultWeight,
    tone,
    "data-id": dataId,
  } = props;

  // Prevent re-renders when context values haven't changed
  const textContextValue = useMemo(
    () => ({
      tone,
      weight,
      variant,
    }),
    [tone, weight, variant]
  );

  const className = useTextStyles({ weight, tone, variant });

  const content = useTruncatedContent({
    children,
    truncate,
  });

  return (
    <TextContext.Provider value={textContextValue}>
      <Box
        data-id={dataId}
        className={className}
        component={component}
        textAlign={align}
      >
        {content}
      </Box>
    </TextContext.Provider>
  );
};

Text.defaultProps = {
  "data-id": "text",
  component: "span",
  variant: defaultVariant,
  tone: "neutral",
  weight: defaultWeight,
  align: "left",
  truncate: false,
};

Text.displayName = "Text";
