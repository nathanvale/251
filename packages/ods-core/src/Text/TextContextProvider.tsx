import * as React from "react";

import {
  TextToneVariants,
  TextVariants,
  TypographyWeightVariants,
} from "@origin-digital/ods-types";

export const TEXT_CONTEXT_DEFAULTS = {
  weight: "regular",
  variant: "body",
} as const;

export interface UseTextProps {
  variant: TextVariants;
  weight: Exclude<TypographyWeightVariants, "bold">;
  tone?: TextToneVariants;
  inline?: boolean;
}

export const TextContext = React.createContext<UseTextProps | false>(false);

export function mergeToContext(
  curContext: UseTextProps | false,
  props: Partial<UseTextProps>,
  defaults: UseTextProps
) {
  let mergedContext;
  if (curContext) {
    mergedContext = {
      tone: props.tone ?? (curContext && curContext.tone) ?? defaults?.tone,
      variant:
        props.variant ?? (curContext && curContext.variant) ?? defaults.variant,
      weight:
        props.weight ?? (curContext && curContext.weight) ?? defaults.weight,
      inline:
        props.inline ?? (curContext && curContext.inline) ?? defaults.inline,
    };
  } else {
    mergedContext = {
      tone: props.tone ?? defaults?.tone,
      variant: props.variant ?? defaults.variant,
      weight: props.weight ?? defaults.weight,
      inline: props.inline ?? defaults.inline,
    };
  }

  return mergedContext;
}

export interface TextContextProviderProps extends Partial<UseTextProps> {
  children?: React.ReactNode;
  defaults?: UseTextProps;
}

export function TextContextProvider({
  variant,
  weight,
  tone,
  inline,
  children,
  defaults = TEXT_CONTEXT_DEFAULTS,
}: TextContextProviderProps) {
  const curContext = React.useContext(TextContext);

  const textContextValue = React.useMemo(() => {
    return mergeToContext(
      curContext,
      {
        tone,
        variant,
        weight,
        inline,
      },
      defaults
    );
  }, [curContext, defaults, tone, variant, weight, inline]);

  return (
    <TextContext.Provider value={textContextValue}>
      {children}
    </TextContext.Provider>
  );
}
