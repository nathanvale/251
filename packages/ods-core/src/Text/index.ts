export { Text } from "./Text";
export { TextContextProvider, TextContext } from "./TextContextProvider";
export { useTextStyles, useTextStylesFromContext } from "./useTextStyles";

export type TextProps = import("./Text").TextProps;
export type UseTextProps = import("./TextContextProvider").UseTextProps;
export type TextContextProviderProps = import("./TextContextProvider").TextContextProviderProps;
