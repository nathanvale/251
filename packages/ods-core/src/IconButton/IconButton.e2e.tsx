import { E2ETests } from "@origin-digital/ods-types";
import { generateIconButton } from "./IconButton.helper";

export const tests: E2ETests = [
  {
    label: "Default IconButton",
    Code: () => generateIconButton(),
  },
  {
    label: "Inherit color from Text",
    Code: () =>
      generateIconButton({ color: "inherit" }, {}, { tone: "positive" }),
  },
  {
    label: "Override button color",
    Code: () => generateIconButton({ color: "primary" }, { tone: "positive" }),
  },
  {
    label: "Disabled",
    Code: () => generateIconButton({ color: "primary", disabled: true }),
  },
  {
    label: "No icon padding",
    Code: () => generateIconButton({ color: "primary", noIconPadding: true }),
  },
];
