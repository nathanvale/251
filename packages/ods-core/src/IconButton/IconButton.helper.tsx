import React from "react";
import { IconBankAccount, SvgIconProps } from "@origin-digital/ods-icons";
import { Text, TextProps } from "../Text/Text";
import { IconButton, IconButtonProps } from "./IconButton";

export const generateIconButton = (
  iconButtonProps: Partial<IconButtonProps> = {},
  iconProps: Partial<SvgIconProps> = {},
  textProps?: TextProps
) => {
  const iconButton = (
    <IconButton {...iconButtonProps}>
      <IconBankAccount {...iconProps} />
    </IconButton>
  );
  if (textProps) {
    return <Text {...textProps}>{iconButton}</Text>;
  }
  return iconButton;
};
