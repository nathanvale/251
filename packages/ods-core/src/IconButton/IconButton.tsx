import React, { useContext, useMemo } from "react";
import clsx from "clsx";
import {
  IconButtonProps as MuiIconButtonProps,
  IconButton as MuiIconButton,
  Theme,
} from "@material-ui/core";
import { makeStyles, fade } from "@material-ui/core/styles";
import {
  ComponentBaseProps,
  IconButtonColor,
  SvgIconToneVariants,
} from "@origin-digital/ods-types";
import { SvgIconContext, SvgIconProps } from "@origin-digital/ods-icons";
import { ButtonActionProps } from "../Button/Button";
import { TextContext, UseTextProps } from "../Text/TextContextProvider";

export interface IconButtonProps extends ComponentBaseProps, ButtonActionProps {
  children: React.ReactElement<SvgIconProps>;
  "aria-label"?: string;
  color?: IconButtonColor;
  muiProps?: MuiIconButtonProps;
  noIconPadding?: boolean;
}

const defaultProps: Required<Pick<IconButtonProps, "color" | "data-id">> = {
  color: "primary",
  "data-id": "icon-button",
};

const useIconButtonStyles = makeStyles((theme: Theme) => ({
  root: {
    verticalAlign: "initial",
    "&:hover": {
      backgroundColor: fade(theme.palette.common.black, 0.08),
    },
  },
  noIconPadding: {
    margin: "-12px",
  },
}));

export const getSvgTone = (
  color: IconButtonColor,
  disabled?: boolean,
  textContext?: false | UseTextProps
): SvgIconToneVariants => {
  if (disabled) {
    return "disabled";
  }
  if (color === "inherit" && textContext && textContext.tone) {
    return textContext.tone;
  }
  return color;
};

export const IconButton = ({
  children,
  color = defaultProps.color,
  noIconPadding,
  disabled,
  ...props
}: IconButtonProps) => {
  const styles = useIconButtonStyles();
  const textContext = useContext(TextContext);
  const tone = useMemo(() => getSvgTone(color, disabled, textContext), [
    color,
    disabled,
    textContext,
  ]);

  return (
    <MuiIconButton
      {...props}
      className={clsx(styles.root, noIconPadding && styles.noIconPadding)}
      disabled={disabled}
      size="medium"
      {...props.muiProps}
    >
      <SvgIconContext.Provider value={tone}>{children}</SvgIconContext.Provider>
    </MuiIconButton>
  );
};

IconButton.displayName = "IconButton";

IconButton.defaultProps = defaultProps;
