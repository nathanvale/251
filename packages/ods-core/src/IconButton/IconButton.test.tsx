import { render } from "@origin-digital/ods-testing-library";
import { UseTextProps } from "../Text/TextContextProvider";
import { generateIconButton } from "./IconButton.helper";
import { getSvgTone } from "./IconButton";

describe("getSvgTone()", () => {
  const getTextContext = (tone: UseTextProps["tone"]): UseTextProps => ({
    variant: "body",
    weight: "regular",
    tone,
  });
  test("should return disabled", () => {
    expect(getSvgTone("primary", true)).toBe("disabled");
  });

  test("should return the text color if set to inherit", () => {
    expect(getSvgTone("inherit", false, getTextContext("positive"))).toBe(
      "positive"
    );
  });

  test("should return the color if there is no text context", () => {
    expect(getSvgTone("primary")).toBe("primary");
  });
});

describe("<IconButton />", () => {
  test("should render", () => {
    expect(render(generateIconButton()).container.firstChild).toMatchSnapshot();
  });

  test("should render with no padding", () => {
    const { container } = render(generateIconButton({ noIconPadding: true }));
    expect(
      container.querySelector("button[class*='noIconPadding']")
    ).not.toBeNull();
  });
});
