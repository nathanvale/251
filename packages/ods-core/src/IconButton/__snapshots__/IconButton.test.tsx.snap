// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton /> should render 1`] = `
<button
  class="MuiButtonBase-root MuiIconButton-root makeStyles-root-1"
  data-id="icon-button"
  tabindex="0"
  type="button"
>
  <span
    class="MuiIconButton-label"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root svg-icon-primary-small-svg-icon-3"
      data-id="IconAccountBalance"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M4 10h3v7H4zM10.5 10h3v7h-3zM2 19h20v3H2zM17 10h3v7h-3zM12 1L2 6v2h20V6z"
      />
    </svg>
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;
